src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.HillClimbing.cs

```diff
@@ -378,9 +394,28 @@ private void ChangeThreadCount(int newThreadCount, StateOrTransition state)
                 _completionsSinceLastChange = 0;
             }
 
-            private void LogTransition(int newThreadCount, double throughput, StateOrTransition state)
+            private void LogTransition(int newThreadCount, double throughput, StateOrTransition stateOrTransition)
             {
-                // TODO: Log transitions
+                // Use the _log array as a circular array for log entries
+                int index = (_logStart + _logSize) % LogCapacity;
+
+                if(_logSize == LogCapacity)
+                {
+                    _logStart = (_logStart + 1) % LogCapacity;
+                    _logSize--; // hide this slot while we update it
+                }
+
+                ref LogEntry entry = ref _log[index];
+
+                entry.tickCount = Environment.TickCount;
+                entry.stateOrTransition = stateOrTransition;
+                entry.newControlSetting = newThreadCount;
+                entry.lastHistoryCount = (int)(Math.Min(_totalSamples, _samplesToMeasure) / _wavePeriod) * _wavePeriod;
```

Comment:
I think it would be fine to cast to int before the divide, as done to calculate the sample count in Update()